### ISSUE TITLE
Warning or error on not universally satisfiable parametricity

### ISSUE THREAD

**Issue by strangemonad** [microsoft/pyright#3753]
**Is your feature request related to a problem? Please describe.**

This might end up impacting the implementation approach of #3684.

I've started trying to document a comparison of some of the consequences of the subtle differences in how pyright and mypy have decided to implement a type system for python and why we chose pyright. As I was documenting small examples, the following caught me by surprise.

Let's start with something something really simple that I thought would get caught by pyright:

```python
def foo(a: Any) -> str:
    return a

reveal_type(foo) # Type of "foo" is "(a: Any) -> str"

s = foo(42)
reveal_type(s) # Type of "s" is "str"
```

Clearly, it's the case that some call-sites can satisfy sum: Any -> str but it's not the case for all values of `a`.  In particular, pyright believes `s` is a `str` but preservation is broken for `foo(42): int <: Any -> str` and  then you can't guarantee progress for anything assuming to use `s` as a `str`.

The generic version (with a type bound) doesn't fair better:

```python
T = TypeVar("T", bound=Any) # Note: this DOES produce the expected error if you leave out the `Any` upper type bound.
def foo(a: T) -> str:
    return a

reveal_type(foo) # Type of "foo" is "(a: T@foo) -> str"

s = foo(42)
reveal_type(s) # Type of "s" is "str"
```

The un-annotated version, `def foo(a) -> str:` *does* fail in strict mode but due to `Unknown` parameter type, not because the types aren't satisfiable.

For what it's worth, I came to the distilled example above from the following slightly more complex example:

```python
def string_sum(a: Any, b: Any) -> str:
   return a + b
```

**Describe the solution you'd like**
I haven't thought about this a whole lot yet in the Python setting and I'm mostly thinking of this coming from other stricter languages.

0. Since I haven't thought about this much in the context of python, this might be completely ill-advised and start failing way too many conventionally runtime-valid programs. Maybe all that's needed is better documentation in the [type inference docs](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md) about how `Any` is treated during bi-directional type checking in function scopes. Certainly, Pierce-style bi-di checking in java or kotlin wouldn't allow something like this so it probably deserves some mention?

That being said, this sort if checking is ideally what I'd like when I switch from basic to strict. Also, for what it's worth, for our use case, I'm happy to have a "fast-path" mode where you get most errors quickly and a slightly slower strict mode.

1. Looking at only the function signature and body, without the hint of some other protocol, warn that `forall a:Any` it's not possible for `foo` to always return `str`. My mental model for this is Girard-style parametricity - we basically know nothing of `a` with just the function signature and body. I can understand why this might be undesirable to solve in the case of `foo(a:Any) -> str` but I expected pyright to raise an error for `foo(a:T <: Any) -> str` because in this case we really should know nothing about `a` (note, however this does work correctly for the unbounded type-var case `foo(a:T) -> str`

2. If you have the function body and a call site, emit an error when call-site inference fails e.g. `foo(42)` fails but `foo("hello")` does not. This approach also has downsides. It requires you to specialize each call-site rather than use call sites to merely inform and unify a single function type (you could probably take the approach of OCaml applicative functors here an only care about call sites up to type equality at the call site?). I believe pyre sort of does something like this since they're interested in tainted dataflows for security analysis. This approach obviously only works if you have the function body. If you only have pyi signatures then you have to trust that it's `(Any) -> str`.

3. In the case of `string_sum` it would be nice to be able to lean on protocols as part of refining the expected types. e.g. the `__add__`-able protocol stipulates (in prose) that the return types should be of the same class hierarchy.
